[MOB-130] - Documentation for New InApp (#39)
Iterable/swift-sdk: README.md

*Swift*
	
```swift
func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([Any]?) -> Void) -> Bool {
	guard let url = userActivity.webpageURL else {
		return false
	}

	// This will track the click, retrieve the original URL and call `handleIterableURL:context:` with the original URL
	return IterableAPI.handle(universalLink: url)
}
```

*Objective-C*
	
```objective-c
- (BOOL)application:(UIApplication *)application continueUserActivity(NSUserActivity *)userActivity restorationHandler:(void (^)(NSArray *restorableObjects))restorationHandler {
	// This will track the click, retrieve the original URL and call `handleIterableURL:context:` with the original URL
	return [IterableAPI handleUniversalLink:userActivity.webpageURL];
}
```
	

> #### In-app notifications

Reviewer: - In-app messages

Reviewer: Thanks! In the future I'd love to help more with things like this, but I need to get spun up on using the SDK before I try to document how to do it :)

*Swift*
	
```swift
func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([Any]?) -> Void) -> Bool {
	guard let url = userActivity.webpageURL else {
		return false
	}

	// This will track the click, retrieve the original URL and call `handleIterableURL:context:` with the original URL
	return IterableAPI.handle(universalLink: url)
}
```

*Objective-C*
	
```objective-c
- (BOOL)application:(UIApplication *)application continueUserActivity(NSUserActivity *)userActivity restorationHandler:(void (^)(NSArray *restorableObjects))restorationHandler {
	// This will track the click, retrieve the original URL and call `handleIterableURL:context:` with the original URL
	return [IterableAPI handleUniversalLink:userActivity.webpageURL];
}
```
	
#### In-app messages

> If you are already using in-app messages with IterableSDK, please checkout the [migration section](#Migrating-in-app-messages-from-previous-version-of-SDK).

Reviewer: ```suggestion
If you are already using in-app messages with IterableSDK, please check out the [migration section](#Migrating-in-app-messages-from-previous-version-of-SDK).
```

	
Custom action is specified by passing `itbl://customActionName` as the link URL (notice the `itbl` scheme name). If the in-app message's `href` property contains custom action name, tapping the  message will call the `handle` method of `IterableConfig.customActionDelegate` (an object that conforms to the `IterableCustomActionDelegate` protocol). If `customActionDelegate` is not set, nothing will happen by default.

Take a look at [this sample code](https://github.com/Iterable/swift-sdk/blob/master/sample-apps/swift-sample-app/swift-sample-app/AppDelegate.swift) that demonstrates how to implement and use the `IterableURLDelegate` and `IterableCustomActionDelegate` protocols.
	
```swift
let config = IterableConfig()
config.urlDelegate = YourCustomUrlDelegate()
config.customActionDelegate = YourCustomActionDelegate()
```
	
##### Changing the display interval between in-app messages

To customize the time delay between successive in-app messages (default value of 30 seconds), set `IterableConfig.inAppDisplayInterval` to an appropriate value (in seconds). 

> ##### Migrating in-app messages from previous version of SDK

Reviewer: ```suggestion
##### Migrating in-app messages from the previous version of the SDK
```

> If you are already using in-app messages, then you will have to make the following changes to your code.

Reviewer: ```suggestion
If you are already using in-app messages, then you will have to make the following changes to your code:
```

> 1. `spawnInAppNotification` is no longer needed. In-app messages will be displayed automatically without needing any code changes. Remove all `spawnInAppNotification` calls.

Reviewer: ```suggestion
1. `spawnInAppNotification` is no longer needed. In-app messages will be displayed automatically. Remove all `spawnInAppNotification` calls.
```

2. Stop polling for in-app messages. The SDK will issue a callback using `config.inAppDelegate` [as explained above](#overriding-whether-to-show-or-skip-a-particular-in-app-message).

> 3. Remove 'IterableAPI.getInAppMessages()' call and use `IterableAPI.inAppManager.getInAppMessages()` instead.

Reviewer: ```suggestion
3. Remove calls to 'IterableAPI.getInAppMessages()' and use `IterableAPI.inAppManager.getInAppMessages()` instead.
```

# Using the SDK

##### 1. Handle Push Notifications

> When the user taps on the push notification or one of the action buttons, the system calls `UNUserNotificationCenterDelegate`'s [userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler:](https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/1649501-usernotificationcenter?language=objc). Pass this call to **`IterableAppIntegration`** to track push open event and perform the associated action (see below for custom action and URL delegates).

Reviewer: - Do you want to link to the Swift-based doc instead?
- Not sure if the `IterableAppIntegration` needs to be bold
- Could you clarify "pass this call to..."?

Author: Eagle eyes.

	
**Swift**
	
```swift
public func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
	IterableAppIntegration.userNotificationCenter(center, didReceive: response, withCompletionHandler: completionHandler)
}
```

**Objective-C**
	
```objective-c
- (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler {
	[IterableAppIntegration userNotificationCenter:center didReceiveNotificationResponse:response withCompletionHandler:completionHandler];
}
```
	

> ##### 2. Deep Linking

Reviewer: - Deep linking

* Handling Links from Push Notifications
	

> Push notifications and action buttons may have `openUrl` actions attached to them. When a URL is specified, the SDK first calls `urlDelegate` specified in your `IterableConfig` object. You can use this delegate to handle `openUrl` actions the same way as you handle normal deep links. If the delegate is not set or if it returns `false` (the default), the SDK will open Safari with that URL. If you want to navigate to a UIViewController on receiving a deep link, you should do so in the `urlDelegate`. 

Reviewer: - calls the `urlDelegate`
- ~UIViewController~ view controller
- ~you should do so~ do so

> In the code below, `DeepLinkHandler` is a custom handler which is reponsible for deep link navigation. You have to provide implementation for deep link navigation. Please see [sample application](https://github.com/Iterable/swift-sdk/blob/master/sample-apps/swift-sample-app?raw=true) for a reference implementation.

Reviewer: - provide the implementation

func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
	...
	// Initialize Iterable API
	let config = IterableConfig()
	...
	config.urlDelegate = self
	IterableAPI.initialize(apiKey: apiKey, launchOptions:launchOptions, config: config)
	...
}

// Iterable URL Delegate. It will be called when you receive 
// an `openUrl` event from push notification.
func handle(iterableURL url: URL, inContext context: IterableActionContext) -> Bool {
	return DeeplinkHandler.handle(url: url)
}
```
	
**Objective-C**
	
```objective-c
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
	...
	// Initialize Iterable SDK
	IterableConfig *config = [[IterableConfig alloc] init];
	...
	config.urlDelegate = self;
	[IterableAPI initializeWithApiKey:@"YOUR API KEY" launchOptions:launchOptions config:config];
	...
}
	
- (BOOL)handleIterableURL:(NSURL *)url context:(IterableActionContext *)context {
	// Assuming you have a DeeplinkHandler class that handles all deep link URLs and navigates to the right place in the app
	return [DeeplinkHandler handleUrl:url];
}
```
		

> * Handling Email Links

Reviewer: - Should this be a sub-header?
- Handling email links

> For Universal Links to work with link rewriting in emails, you need to set up apple-app-site-association file in the Iterable project. More instructions here: [Setting up iOS Universal Links](https://support.iterable.com/hc/en-us/articles/115000440206-Setting-up-iOS-Universal-Links).

Reviewer: - universal links
- ~you need to~
- set up an
- Suggestion:

> For universal links to work with link rewriting in emails, [set up an **apple-app-site-association** file](https://support.iterable.com/hc/articles/115000440206-Setting-up-iOS-Universal-Links) in the Iterable project.

> When an email link is clicked your `UIApplicationDelegate`'s [application:continueUserActivity:restorationHandler:](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623072-application?language=swift) method is called. If you already have an Iterable `urlDelegate` defined (see *Handling Links from Push Notifications* section above), the same handler can be used for email deep links by calling `handleUniversalLink:`.

Reviewer: - is clicked,
- Actually link up to the prior section with an anchor link.
- Swift references instead of Objective-C references?

Author: Not sure I understand "link up to prior section...". Also, the reference here is Swift not ObjC.

**Swift**
	
```swift
func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([Any]?) -> Void) -> Bool {
	guard let url = userActivity.webpageURL else {
		return false
	}

	// This will track the click, retrieve the original URL and call `handleIterableURL:context:` with the original URL
	return IterableAPI.handle(universalLink: url)
}
```

**Objective-C**
	
```objective-c
- (BOOL)application:(UIApplication *)application continueUserActivity(NSUserActivity *)userActivity restorationHandler:(void (^)(NSArray *restorableObjects))restorationHandler {
	// This will track the click, retrieve the original URL and call `handleIterableURL:context:` with the original URL
	return [IterableAPI handleUniversalLink:userActivity.webpageURL];
}
```
	

> Alternatively, call `getAndTrackDeeplink` along with a callback to handle the original deeplink url. You can use this method for any incoming URLs, as it will execute the callback without changing the URL for non-Iterable URLs.

Reviewer: - ~deeplink~ deep link

**Swift**
	
```swift
func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([Any]?) -> Void) -> Bool {
	guard let url = userActivity.webpageURL else {
		return false
	}

	IterableAPI.getAndTrack(deeplink: url) { (originalUrl) in
		// Handle original url deeplink here
	}
	return true
}
```

**Objective-C**
	
```objective-c
- (BOOL)application:(UIApplication *)application
		continueUserActivity(NSUserActivity *)userActivity 
		restorationHandler:(void (^)(NSArray *restorableObjects))restorationHandler {

	[IterableAPI getAndTrackDeeplink:iterableLink callbackBlock:^(NSString* originalURL) {
		//Handle Original URL deeplink here
	}];

	return true;
}
```
	
##### 3. In-App Notifications

###### Default behavior

> By default, when an in-app message arrives from the server, the SDK automatically shows it if the app is in the foreground. If an in-app message is already showing when the new message arrives, the new in-app message will be shown 30 seconds after the currently displayed in-app message closes (see how to change this default value below). Once an in-app message is shown, it will be "consumed" from the server queue and removed from the local queue as well. There is no need to write any code to get this default behavior. 

Reviewer: - maybe use an anchor link to jump to the below section about how to change the value of the delay

	
**Swift**
	
```swift
// Get the in-app messages list
let messages = IterableAPI.inAppManager.getMessages()
	
// Show an in-app message 
IterableAPI.inAppManager.show(message: message)
	
// Show an in-app message without consuming, i.e., not removing it from the queue
IterableAPI.inAppManager.show(message: message, consume: false)
	
```	
	
**Objective-C**
	
```objc
// Get the in-app messages list
NSArray *messages = [IterableAPI.inAppManager getMessages];
	
// Show an in-app message 
[IterableAPI.inAppManager showMessage:message];
	
// Show an in-app message without consuming, i.e., not removing it from the queue
[IterableAPI.inAppManager showMessage:message consume:NO callbackBlock:nil];
	
```	

###### When user clicks a button/link in the in-app message
If the in-app message's clicked button/link's `href` contains a URL (which is usually the case), tapping that button/link will call the `handle` method of `IterableConfig.urlDelegate` (an object of type `IterableURLDelegate`), if one is set. By default if this delegate is not set, tapping the button/link will open Safari with the `href` of the clicked button/link.
	

> If the in-app message's `href` property contains custom action name (instead of a URL), tapping the  message will call the `handle` method of `IterableConfig.customActionDelegate` (an object that conforms to the `IterableCustomActionDelegate` protocol). If `customActionDelegate` is not set, nothing will happen by default.

Reviewer 2: Custom action is specified by passing `itbl://customActionName` as the link URL, we need to clarify that.

Take a look at [this sample code](https://github.com/Iterable/swift-sdk/blob/master/sample-apps/swift-sample-app/swift-sample-app/AppDelegate.swift) that demonstrates how to implement and use the `IterableURLDelegate` and `IterableCustomActionDelegate` protocols.
	
```swift
let config = IterableConfig()
config.urlDelegate = YourCustomUrlDelegate()
config.customActionDelegate = YourCustomActionDelegate()
```
	
###### Changing the display interval between in-app messages
To customize the time delay between successive in-app messages (default value of 30 seconds), set `IterableConfig.inAppDisplayInterval` to an appropriate value (in seconds). 

> ##### 4. Tracking Custom Events

Reviewer: - Tracking custom events

Custom events can be tracked using `IterableAPI.track(event:...)` calls.
	

> ##### 5. Updating User Fields

Reviewer: Updating user fields

User fields can be modified using `IterableAPI.updateUser` call. You also have `updateEmail` and `updateSubscriptions` methods.
	

> ##### 6. Disabling Push Notifications to a Device

Reviewer: - Disabling push notifications to a device

When a user logs out, you typically want to disable push notifications to that user/device. This can be accomplished by calling `disableDeviceForCurrentUser`. Please note that it will only attempt to disable the device if you have previously called `registerToken`.
	
In order to re-enable push notifcations to that device, simply call `registerToken` as usual when the user logs back in.
	

> ##### 7. Uninstall Tracking

Reviewer: - Uninstall tracking

Iterable will track uninstalls with no additional work by you. 

> This is implemented by sending a second push notification some time (currently, twelve hours) after the original campaign. If we receive feedback that the device's token is no longer valid, we assign an uninstall to the device, attributing it to the most recent campaign within twelve hours. A "real" campaign send (as opposed to the later "ghost" send) can also trigger recording an uninstall. In this case, if there was no previous campaign within the attribution period, an uninstall will still be tracked, but it will not be attributed to any campaign.

Reviewer: - ~we receive~ the app receives
- ~we assign~ it assigns
- can also trigger the 

Suggestion:

> To do this, Iterable sends a silent push notification some time (currently, twelve hours) after a campaign has been sent. Based on this silent push notification, if Iterable receives feedback that the device token is no longer valid, it assigns an uninstall to the device based on the prior campaign. Similarly, if a "real" campaign uncovers an invalid device token, it will also check for a prior (within twelve hours) campaign to mark as the cause for the uninstall. If there was no recent campaign, Iterable still tracks the uninstall, but does not attribute it to a campaign.

		```
		
3. ##### InApp Notifications
	###### Default Behavior
	By default, when an inApp message arrives from the server it is automatically shown by the SDK provided the app is in foreground. If an inApp message is already showing when the message arrives, the new inApp will be shown 30 seconds (see how to change this default value below) after the currently displaying inApp is closed. Once an inApp message is shown, it will be "consumed" from the server queue and removed from the local queue as well. There is no need to write any code to get this default behavior. 
	

> 	###### Overriding Whether to Show or Skip a Particular InApp Message

Reviewer: - No need to indent
- Capitalization: Overriding whether to show or skip a particular in-app message

> 	When an inApp message arrives from the server, the `onNew` method of `IterableInAppDelegate` is called. This delegate is set via the `inAppDelegate` property of `IterableConfig`. You can set `IterableConfig.inAppDelegate` to a custom class to override the default behavior. This class just needs to implement the `onNew` method. The `onNew` method should return `.show` to show the message or `.skip` to not show the message at this time.

Reviewer: - No need to indent
- ~inApp~ in-app
- This class ~just needs to~ must implement the `onNew` method

Suggested changes:

> An incoming in-app message triggers a call to the `onNew` method of `IterableConfig.inAppDelegate` (an object of type `IterableInAppDelegate`). To override the default behavior, set `IterableConfig.inAppDelegate` to a custom class that overrides the `onNew` method. `onNew` should return `.show` to show the incoming in-app message or `.skip` to skip showing it.

> 	Swift:

Reviewer: - No need to indent this whole section.
- Maybe make Swift a subheader (no colon), or just bold

> 	```

Reviewer: - should be: ` ```swift `

	class YourCustomInAppDelegate : IterableInAppDelegate {
		func onNew(message: IterableInAppMessage) -> InAppShowResponse {
			// perform custom processing

			// ...
			
			return .show // or .skip
		}
	}
	
	// ...
	
	let config = IterableConfig()
	config.inAppDelegate = YourCustomInAppDelegate()
	IterableAPI.initialize(apiKey: "YOUR API KEY",  launchOptions: nil, config: config)
	```

> 	Objective-C:

Reviewer: - Make this a subheader without a colon (or bold)

> 	```

Reviewer: - Should be: ` ```objc `

Author: Not sure I understand.

Reviewer: Sorry, this should have been two things:

- I'd make "Objective-C" a sub-header or bold
- With the code fence, if you write it as ` ```objc `, then GitHub can syntax-highlight appropriately (since it knows it's Objective-c).

For example:

```
@implementation classname
+ (return_type)classMethod
{
 // implementation
}
- (return_type)instanceMethod
{
 // implementation
}
@end
```

vs

```objc
@implementation classname
+ (return_type)classMethod
{
 // implementation
}
- (return_type)instanceMethod
{
 // implementation
}
@end
```

Author: Great point.

	// Implement this method in your custom class that implements IterableInAppDelegate
	// This will most likely be the global AppDelegate class.
	- (enum InAppShowResponse)onNewMessage:(IterableInAppMessage * _Nonnull)message {
   		// perform custom processing
   		
   		// ...
   		
    	return InAppShowResponseShow; // or InAppShowResponseSkip
	}

	// ...
	
	// Now set this custom class in IterableConfig
	IterableConfig *config = [[IterableConfig alloc] init];
	config.inAppDelegate = self; // or other class implementing the protocol
	[IterableAPI initializeWithApiKey:@"YOUR API KEY" launchOptions:launchOptions config:config];
	```

> 	###### Getting the Local Queue of InApp Messages

Reviewer: - No indent
- Capitalization: Getting the local queue of in-app messages

> 	All inApp messages that arrive from the server are stored in a local queue until they are consumed. IterableSDK exposes `InAppManager` protocol via the `IterableAPI.inAppManager` read-only property to get to locally stored inApp messages. Please be aware that all inApp messages that are shown will be consumed and removed from this queue by default. So you will have to override the default behavior as mentioned above to keep inApp messages around even after they are shown.

Reviewer: Suggestions:

> Until they are consumed by the app, all in-app messages that arrive from the server are stored in a local queue. To access that local queue, use the read-only `IterableAPI.inAppManager` property (which conforms to the `InAppManager` protocol). By default, all in-app messages in the local queue will be consumed and removed from this queue. To keep in-app messages around after they are shown, override the default behavior (as described above).

> 	Swift:

Reviewer: - No indent
- Subheader

> 	```

Reviewer: - ` ```swift `

> 	// Get the inApp messages list

Reviewer: - in-app

	let messages = IterableAPI.inAppManager.getMessages()
	

> 	// Show an inApp message 

Reviewer: - in-app

	IterableAPI.inAppManager.show(message: message)
	

> 	// Show an inApp message without consuming, i.e., not removing it from the queue

Reviewer: - in-app

	IterableAPI.inAppManager.show(message: message, consume: false)
	
	```	
	

> 	Objective-C:

Reviewer: - No indent
- Subheader or bold

> 	```

Reviewer: - ` ```objc `

> 	// Get the inApp messages list

Reviewer: - in-app

    NSArray *messages = [IterableAPI.inAppManager getMessages];
	
	// Show an inApp message 
	[IterableAPI.inAppManager showMessage:message];
	
	// Show an inApp message without consuming, i.e., not removing it from the queue
	[IterableAPI.inAppManager showMessage:message consume:NO callbackBlock:nil];
	
	```	

> 	###### When User Clicks a Button in the InApp Message

Reviewer: - No indent necessary
- Capitalization: When user clicks a button in the in-app message

> 	If the clicked `href` in the inApp message is a url (which is the case most of the time), `IterableURLDelegate` is called. If you don't set a custom class for `IterableConfig.urlDelegate` then mobile Safari will be opened with the clicked href url.

Reviewer: Some suggestions (not sure if I'm totally understanding this, so this may be wrong):

> If the in-app message's `href` contains a URL (which is usually the case), tapping the message will call the `handle` method of `IterableConfig.urlDelegate` (an object of type `IterableURLDelegate`)—which will open Safari to the specified URL. Customize this behavior by setting `IterableConfig.urlDelegate` an instance of `IterableURLDelegate` with a custom `handle` method.

Does the user have to set `IterableConfig.urlDelegate` to something? Or if they leave it blank does Safari open? Not clear on that point.

> 	If the clicked `href` is not a url but a custom action name,  `IterableCustomActionDelegate` is called. Please see example in sample code [here](https://github.com/Iterable/swift-sdk/blob/master/sample-apps/swift-sample-app/swift-sample-app/AppDelegate.swift) to see how to implement `IterableURLDelegate` and `IterableCustomActionDelegate` and handle clicked urls.

Reviewer: - ~url~ URL

Suggestion:

> If the in-app message's `href` property contains custom action name (instead of a URL), tapping the  message will call the `handle` method of `IterableConfig.customActionDelegate` (an object that conforms to the `IterableCustomActionDelegate` protocol).
>
> Take a look at [this sample code](https://github.com/Iterable/swift-sdk/blob/master/sample-apps/swift-sample-app/swift-sample-app/AppDelegate.swift) that demonstrates how to implement and use the `IterableURLDelegate` and `IterableCustomActionDelegate` protocols.

Does the user have to set IterableConfig.customActionDelegate to something? Or if they leave it blank is there default behavior?

Author: There is no default behavior in this case, since we don't really know what to do with custom actions.

> 	```

Reviewer: - ` ```swift `

	let config = IterableConfig()
	config.urlDelegate = YourCustomUrlDelegate()
	config.customActionDelegate = YourCustomActionDelegate()
	```
	

> 	###### Changing the Display Interval Between InApp Messages

Reviewer: - No indent
- Capitalization: Changing the display interval between in-app messages

> 	If you want to change the display interval, i.e., the time delay to show two successive inApp messages, to some value other than 30 seconds, change `IterableConfig.inAppDisplayInterval` to the appropriate value.

Reviewer: - No indent

Suggestion:

> To customize the time delay between successive in-app messages (default value of 30 seconds), set `IterableConfig.inAppDisplayInterval` to an appropriate value.

Probably good to specify units here—seconds? milliseconds?

> 4. ##### Tracking Custom Events

Reviewer: - Shouldn't number and half a Markdown header
- Capitalization: Tracking custom events

Author: Tracking custom events is another topic (not in-app). I will take a look at numbering and indenting throughout the entire file. That should fix it.
